api/lib/flight_job_script_api/subprocess.rb

```diff
@@ -0,0 +1,191 @@
+#==============================================================================
+# Copyright (C) 2021-present Alces Flight Ltd.
+#
+# This file is part of Flight Job Script Service.
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# Flight Job Script Service is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with Flight Job Script Service. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on Flight Job Script Service, please visit:
+# https://github.com/openflighthpc/flight-job-script-service
+#==============================================================================
+
+module FlightJobScriptAPI
+  class Subprocess
+    class Result < Struct.new(:stdout, :stderr, :exitstatus, :pid); end
+
+    def initialize(dir:, env:, gid:, logger:, timeout:, uid:)
```

Comment:
It would be more useful if we provided the `user` instead of the `uid`/`gid`. I can't see a situation where we would want to run with arbitrary `uid`/`gid` as oppose to the a "user".

Reply:
For now I think this is good enough.  To have `Subprocess` work for file manager api and desktop restapi the group handling needs to be reworked.  They both require the `gid` of the existing process (`root`) to be removed, and for both the primary and supplementary groups to be added.  So, this isn't quite right and the correct solutions is probably to take a user/username/uid along with a `supplementary_groups` option.

Reply:
Ah far point, maybe best to take this class and port it into `file-manager-api` and see what changes are required?

Hopefully that will result in a fairly feature complete `Subprocess` class.

Reply:
OK. You convinced me. https://github.com/openflighthpc/flight-job-script-service/pull/41/commits/8dce3fdf21044634364661fff37985b787d45a72

Reply:
:+1: